### PR TITLE
Full package review in prep of CRAN release

### PR DESCRIPTION
**Pull request by bahadzie**

### CODE REVIEW

File: .Rbuildignore [hunk 1]
```diff
@@ -9,3 +9,5 @@
 ^.*\.Rproj$
 ^\.Rproj\.user$
 ^CITATION\.cff$
+^__.*$
+^\.vscode$
```
Comment (chartgerink) [inline .Rbuildignore:13]: I don't think this will match the file you're trying to explicitly ignore as the `$` indicates the path should also end on `.vscode`. See an example here: https://regex101.com/r/YwjJpv/1

```suggestion
^\\.vscode/settings\\.json$
```

---

For what it's worth, the usethis help states:

> ‘use_build_ignore()’ is designed to ignore _individual_ files.

File: DESCRIPTION [hunk 2]
```diff
@@ -1,31 +1,36 @@
-Package: packagetemplate
-Title: Your Package Title in Titlecase
+Package: numberize
+Title: Convert words to numbers in multiple languages
```
Comment (chartgerink) [inline DESCRIPTION:2]: CRAN wants title case.

```suggestion
Title: Convert Words to Numbers in Multiple Languages
```

File: DESCRIPTION [hunk 3]
```diff
@@ -1,31 +1,36 @@
...
+           comment = c(ORCID = "0000-0002-3892-6680")),
+    person("Pratik", "Gupte", , "pratik.gupte@lshtm.ac.uk", role = "aut",
+           comment = c(ORCID = "0000-0001-5294-7819")),
+    person("Joshua W.", "Lambert", , "joshua.lambert@lshtm.ac.uk", role = "aut",
            comment = c(ORCID = "0000-0001-5218-3046")),
-    person(given = "Hugo", family = "Gruson", email = "hugo.gruson+R@normalesup.org", role = c("aut"),
-           comment = c(ORCID = "https://orcid.org/0000-0002-4094-1476"))
+    person("Hugo", "Gruson", , "hugo.gruson+R@normalesup.org", role = "aut",
+           comment = c(ORCID = "0000-0002-4094-1476")),
+    person("Banky", "Ahadzie", , "bahadzie@gmail.com", role = c("aut", "cre"))
```
Comment (chartgerink) [inline DESCRIPTION:16]: Do you have an ORCID @bahadzie? We try to use them throughout, so if you have one we can add it 😊

File: DESCRIPTION [hunk 4]
```diff
@@ -1,31 +1,36 @@
...
   )
-Description: Your package description. It must end with a period (".") and
-    include relevant bibliographical references if applicable, using the 
-    following format: Author et al. (2023) <doi:10.5281/zenodo.6619350>.
+Description: Converts numbers written as English, French or Spanish words
+    to their equivalent number.  English, French or Spanish words to their
+    equivalent numeric.
```
Comment (chartgerink) [inline DESCRIPTION:20]: This description seems a bit repetitive. Might this be a good alternative formulation?

```suggestion
Description: Converts written out numbers into their equivalent numbers. Supports numbers written out in English, French, or Spanish.
```

File: LICENSE [hunk 5]
```diff
@@ -1,2 +1,2 @@
 YEAR: 2022
-COPYRIGHT HOLDER: readepi authors
+COPYRIGHT HOLDER: numberizeR authors
```
Comment (chartgerink) [inline LICENSE:2]:
```suggestion
COPYRIGHT HOLDER: numberize authors
```

Not sure why the R is added here?

File: LICENSE [hunk 6]
```diff
@@ -1,2 +1,2 @@
 YEAR: 2022
```
Comment (chartgerink) [inline LICENSE:1]:
```suggestion
YEAR: 2024
```

File: LICENSE.md [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # MIT License
 
-Copyright (c) 2022 readepi authors
+Copyright (c) 2022 numberizeR authors
```
Comment (chartgerink) [inline LICENSE.md:3]:
```suggestion
Copyright (c) 2024 numberize authors
```

File: tests/spelling.R [hunk 8]
```diff
@@ -2,6 +2,6 @@ if (requireNamespace("spelling", quietly = TRUE)) {
   spelling::spell_check_test(
     vignettes = TRUE,
     error = TRUE,
-    skip_on_cran = TRUE
+    skip_on_cran = FALSE
```
Comment (chartgerink) [inline tests/spelling.R:5]: For clarification: Could you expand why you changed this default from the **packagetemplate**?

File: R/numberize.R [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+#' # convert to numbers a scalar
+#' numberize("five hundred and thirty eight")
+#'
+#' # convert a vector of values
+#' numberize(c("dix", "soixante-cinq", "deux mille vingt-quatre"), lang = "fr")
+#'
+#' @export
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+numberize <- function(text, lang = c("en", "fr", "es")) {
+  lang <- match.arg(lang)
```
Comment (chartgerink) [inline R/numberize.R:163]:
```suggestion
numberize <- function(text, lang = c("en", "fr", "es")) {
  lang <- tolower(lang)
  lang <- match.arg(lang)
```

This hardens the input for the `lang` argument - it can happen that `EN` is entered rather easily.

File: tests/testthat/test-numberize.R
Comment (chartgerink) [inline tests/testthat/test-numberize.R:1]: Tests are currently focused on primarily checking the functions work where we expect they would. The test do not observe failures at this time - what are some situations that we can test for where we know the function should not work?

File: README.Rmd [hunk 10]
```diff
@@ -13,56 +13,85 @@ output: github_document
...
 [![R-CMD-check](https://github.com/{{ gh_repo }}/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/{{ gh_repo }}/actions/workflows/R-CMD-check.yaml)
 [![Codecov test coverage](https://codecov.io/gh/{{ gh_repo }}/branch/main/graph/badge.svg)](https://app.codecov.io/gh/{{ gh_repo }}?branch=main)
-[![lifecycle-concept](https://raw.githubusercontent.com/reconverse/reconverse.github.io/master/images/badge-concept.svg)](https://www.reconverse.org/lifecycle.html#concept) 
+[![lifecycle-{{ recon }}](https://www.reconverse.org/images/badge-{{ recon }}.svg)](https://www.reconverse.org/lifecycle.html#{{ recon }})
+[![CRAN status](https://www.r-pkg.org/badges/version/{{ packagename }})](https://CRAN.R-project.org/package={{ packagename }})
+
 <!-- badges: end -->
 
-{{ packagename }} provides functions to ....
+_{{ packagename }}_ is an R package to convert numbers written as English, French or Spanish words from `"zero"` to `"nine hundred and ninety nine trillion, nine hundred and ninety nine billion, nine hundred and ninety nine million, nine hundred and ninety nine thousand, nine hundred and ninety nine"` from a character string to a numeric value.
```
Comment (chartgerink) [inline README.Rmd:32]:
```suggestion
_{{ packagename }}_ is an R package to convert numbers written as English, French or Spanish words from `"zero"` to `"nine hundred and ninety nine trillion, nine hundred and ninety nine billion, nine hundred and ninety nine million, nine hundred and ninety nine thousand, nine hundred and ninety nine"` from a character string to a numeric value.

The package only supports conversion of positive integer values.
```

I was playing around with the function and it was not clear to me it only works for positive integer values. Adding this in here as a suggestion and wonder whether there's any additional place where this can be added explicitly?

File: DESCRIPTION [hunk 11]
```diff
@@ -1,31 +1,36 @@
-Package: packagetemplate
-Title: Your Package Title in Titlecase
+Package: numberize
+Title: Convert words to numbers in multiple languages
 Version: 0.0.1
```
Comment (chartgerink) [inline DESCRIPTION:3]: Is this the version you want to release on? The largest part of this decision depends on how stable this will be. I hope we're committing to making this stable, and as such I would personally recommend going to v1.0.0 for the full release to mark it as such.

**Review by chartgerink — APPROVED** [review]
Thanks @bahadzie for the opportunity to review! 

I love the fact that the scope of the package is very clearly defined, and it delivers on exactly that. Good job 🙌 It made contributing this review a pleasure.

I left some comments, and I would encourage you to especially look at `DESCRIPTION` and `.Rbuildignore` before submitting to CRAN. This in order to prevent the package being rejected on some arbitrary technical ground. I am approving the package regardless, because these things are so minor! 😊 

Feel free to ping me anytime if you want to further discuss something 💜

File: DESCRIPTION [hunk 12]
```diff
@@ -1,31 +1,36 @@
-Package: packagetemplate
-Title: Your Package Title in Titlecase
+Package: numberize
```
Comment (joshwlambert) [inline DESCRIPTION:1]: Running `desc::desc_normalize()` changed the `DESCRIPTION` slightly. It can be useful to standardise the `DESCRIPTION`.

File: R/numberize.R [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+#' @examples
+#' # convert to numbers a scalar
+#' numberize("five hundred and thirty eight")
+#'
+#' # convert a vector of values
+#' numberize(c("dix", "soixante-cinq", "deux mille vingt-quatre"), lang = "fr")
+#'
+#' @export
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+numberize <- function(text, lang = c("en", "fr", "es")) {
```
Comment (joshwlambert) [inline R/numberize.R:162]: Why is the default behaviour for a string with two numbers that are whitespace separated to add them together? I've looked into the function internals to see this is how the function should work, but it can lead to possible mistakes as shown below.
``` r
library(numberize)
numberize::numberize("one one")
#> [1] 2
```

<sup>Created on 2024-04-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

It would be useful if this was documented somewhere.

File: R/numberize.R [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+  lang <- match.arg(lang)
+  vapply(
+    text,
+    .numberize,
+    FUN.VALUE = double(1),
+    lang = lang,
+    USE.NAMES = FALSE
+  )
+}
```
Comment (joshwlambert) [inline R/numberize.R:171]: The behaviour of `numberize::numberize()` to some of R's reserved words are:
``` r
library(numberize)
numberize::numberize(NaN)
#> [1] NA
numberize::numberize(TRUE)
#> [1] 1
numberize::numberize(FALSE)
#> [1] 0
numberize::numberize(NULL)
#> numeric(0)
numberize::numberize(Inf)
#> [1] Inf
```

<sup>Created on 2024-04-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>
These all seem to produce sensible output, but just want to post this in case you wanted to change any of these to maybe return an error or warning.

**Review by joshwlambert — APPROVED** [review]
@bahadzie Nice work putting this package together! 

I agree with @chartgerink that the scope is very well defined and as a result there is a minimal complexity and no dependencies, while still providing really useful functionality.

I've left a few comments throughout the diffs. A few extra comments that cannot be added to the diffs as the files have not changed since the repository was created:

- Update CI workflows to use latest versions from {packagetemplate}. This will use `actions/checkout@v4` and remove the deprecation notices from your GitHub actions
- There is no design principles vignette. This does not need to be added for this release, but in the future it could be useful, especially if the scope of the package expands or takes on a new maintainer
- A Get Started vignette (`packagename.Rmd`) would be useful. Even though the package is simple and has a single exported function having a vignette for the pkgdown website would, in my opinion, by beneficial. Again I do not believe this needs to be done for this release

Happy to discuss anything I've raised if something is unclear.

File: tests/spelling.R (hunk 8 above)
Comment (bahadzie) [reply tests/spelling.R:5]: It's been so long that I don't even remember why.

Before I change it back to the default, is there a reason `skip_on_cran = TRUE` is the default?

File: tests/testthat/test-numberize.R
Comment (bahadzie) [reply tests/testthat/test-numberize.R:1]: I couldn't come up with any. Although some of the tests (listed below) were added as I encountered problems during development

- `test_that("text with non digit word returns NA", {})`
- `test_that("vector with number and words and NA is properly handled", {})`
- `test_that("text with leading and trailing whitespace works", {})`

Happy to add more tests if anyone comes up with any or if I encounter edge conditions during future development.

File: README.Rmd (hunk 10 above)
Comment (bahadzie) [reply README.Rmd:32]: Noted. I've rewritten it to incorporate your suggestion and remove the initial description which isn't specific that the output isn't positive integers. 

My assumption here is that because it is cumbersome to spell out numbers, I didn't expect to see "minus thirty seven point three degrees Celcius" so I based it on what felt natural (positive integers) but I agree its important to be explicit.

File: R/numberize.R (hunk 9 above)
Comment (bahadzie) [reply R/numberize.R:163]: Noted and modified accordingly.

File: R/numberize.R (hunk 13 above)
Comment (bahadzie) [reply R/numberize.R:162]: My initial naive implementation made wrong assumptions. Thanks for spotting this. 

Since it's a `{cleanepi}` dependency and we're trying to submit to CRAN in the short term, I will do what @jamesmbaazam suggested and return NA with a warning for numbers that can be ambiguously interpreted while I collect more feedback.

Eventually, I think this will lead to an additional parameter to control the output e.g. `numberize::numberize("one one", as_vector="TRUE")` will return `c(1, 1)`

I've expanded the test to include ambiguously worded number text.

File: R/numberize.R (hunk 14 above)
Comment (bahadzie) [reply R/numberize.R:171]: This is very helpful. I've updated the code to return NA for all the above situations and included test to confirm it's being enforced.

File: R/numberize.R [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+      return(NA)
+    }
+    number_from(digits)
+  }
+}
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#' Convert a vector string of spelled numbers in a supported language to
+#' its numeric equivalent.
+#'
```
[the remark this answers is not in the document]
Comment (bahadzie) [reply R/numberize.R:146]: It's actually the first paragraph of README.md 

> *numberize* is an R package to convert numbers written as English,
> French or Spanish words from `"zero"` to `"nine hundred and ninety nine
> trillion, nine hundred and ninety nine billion, nine hundred and ninety
> nine million, nine hundred and ninety nine thousand, nine hundred and
> ninety nine"` from a character string to a numeric value.

It makes sense to include it in the Roxygen labels as well because then it will show for `?numberize`

File: LICENSE.md (hunk 7 above)
Comment (bahadzie) [reply LICENSE.md:3]: :+1:

File: LICENSE (hunk 6 above)
Comment (bahadzie) [reply LICENSE:1]: :+1:

File: LICENSE (hunk 5 above)
Comment (bahadzie) [reply LICENSE:2]: Was initially called `numberizeR` one to many Rs if you ask me

File: DESCRIPTION (hunk 2 above)
Comment (bahadzie) [reply DESCRIPTION:2]: :+1:

File: DESCRIPTION [hunk 16]
```diff
@@ -1,31 +1,36 @@
...
-           comment = c(ORCID = "https://orcid.org/0000-0001-5294-7819")),
-    person(given = "Joshua W.", family = "Lambert", email = "joshua.lambert@lshtm.ac.uk", role = c("aut"),
+    person("Avinash", "Laddha", , "avinash@data.org", role = "aut"),
+    person("James M.", "Azam", , "james.azam@lshtm.ac.uk", role = "aut",
+           comment = c(ORCID = "0000-0001-5782-7330")),
+    person("Jaime A.", "Pavlich-Mariscal", , "jpavlich@javeriana.edu.co", role = "ctb",
+           comment = c(ORCID = "0000-0002-3892-6680")),
+    person("Pratik", "Gupte", , "pratik.gupte@lshtm.ac.uk", role = "aut",
+           comment = c(ORCID = "0000-0001-5294-7819")),
+    person("Joshua W.", "Lambert", , "joshua.lambert@lshtm.ac.uk", role = "aut",
```
[the remark this answers is not in the document]
Comment (bahadzie) [reply DESCRIPTION:12]: Can I list everyone who took the time to review the package as contributors or is it only accepted PR submitters that are considered contributors?

File: DESCRIPTION (hunk 3 above)
Comment (bahadzie) [reply DESCRIPTION:16]: No I don't. Do I have to? If not I'm happy to skip it. I'm trying to reduce the number of things I'm signed up to.

File: DESCRIPTION (hunk 11 above)
Comment (bahadzie) [reply DESCRIPTION:3]: I believe as epiverse we are committed to stable packages. I assumed that epiverse packages always start life on CRAN at version 0.x.y (e.g. {cfr}, {finalsize}) and only get to 1.x.y if there are breaking changes between v0 and v1. Out of interest is there a metric for release stability?

File: .Rbuildignore (hunk 1 above)
Comment (bahadzie) [reply .Rbuildignore:13]: I'm actually trying to exclude a folder. I think I've nailed it now. Thanks

File: tests/testthat/test-numberize.R
Comment (bahadzie) [reply tests/testthat/test-numberize.R:1]: Additional tests added thanks to @joshwlambert for spotting some rough edges.

```r
test_that("text with ambigious number conversion returns NA", {})
test_that("NA to return NA", {})
test_that("NaN to return NA", {})
test_that("TRUE to return NA", {})
test_that("FALSE to return NA", {})
test_that("NULL to return NA", {})
test_that("Inf to return NA", {})
```

File: DESCRIPTION (hunk 16 above)
[the remark this answers is not in the document]
Comment (joshwlambert) [reply DESCRIPTION]: Reviewers can be acknowledged in the description with the `role = "rev"`. See for example https://github.com/epiverse-trace/epiparameter/blob/main/DESCRIPTION

File: DESCRIPTION (hunk 12 above)
Comment (bahadzie) [reply DESCRIPTION:1]: Done

File: DESCRIPTION (hunk 11 above)
Comment (chartgerink) [reply DESCRIPTION:3]: I agree that we are committed to stable packages. For semantic versioning, the default interpretation is that anything <1.0.0 is considered unstable (see https://semver.org/).

If there is a deviation from that standard, that is of course okay, but I am unaware of it being explicit within Epiverse.

Comment (bahadzie) [reply DESCRIPTION:3]: Spot on. #5 [Semantic Versioning Specification (SemVer)](https://semver.org/#semantic-versioning-specification-semver)

I will release as 1.0.0. I believe this package is stable. Having a narrow scope helps.
